course02/homework01/Tomasz/list.h

```diff
@@ -0,0 +1,480 @@
+#pragma once
+
+#include <cstddef>
+#include <cassert>
+
+template<typename T>
+class List
+{
+private:
+	struct Node
+	{
+		Node(const T& data);
+
+		T data;
+
+		Node* next;
+		Node* prev;
+	};
+
+public:
+	template<typename E>
```

Comment:
I'm curious why an iterator needs a template here?

Reply:
I'm using it to create const and mutable iterators without defining them separately. So the const iterator is described as Iterator<const T> and standard iterator is simply Iterator<T>. I probably might have also just used a List<const T>::Iterator as a const iterators and then I wouldn't have to define the template here. I don't know what would be the cleaner solution.

Reply:
Makes a lot of sense now, cheers!

Reply:
Np. 
I've just checked and:
- indeed if I use the same plain iterator for const and mutable purposes then I'm able to edit list elements even if the list itself is conts (which is undesirable),
- in fact I could not have used `List<const T>::Iterator` as the `owner` of the iterator is still simply `List<T>` and node is `List<T>::Node` so there is a constructor parameter mismatch.

Reply:
Thanks! I didn't know that there were such requirements for iterators to have iterator_category etc. Should the `value_type` of `const iterator` be `const T` or simply `T`, though? I've implemented it as T, but I'm not sure if that's correct.

I have also fixed the problem with performing operations between `const iterator` and `iterator` by deriving the second from the first.

Reply:
I've actually started implementing it with unique_pointers at first but the fact that I had to treat `prev` pointers and `next` pointers in nodes differently seemd to me as a code pollution. But I'm aware that this list is a rather specific example and overall it's a better idea to use smart pointers :)